src/Build/Collections/PropertyDictionary.cs

```diff
@@ -241,21 +241,24 @@ public IEnumerator<T> GetEnumerator()
         {
             lock (_properties)
             {
-                return _properties.Values.GetEnumerator();
+                var snapshot = new List<T>(_properties.Values);
```

Comment:
Generating an iterator under the lock (`foreach + yield return`) will also hold the lock until the enumeration is done + avoid the snapshot allocations, which I expect could be significant for areas like `Lookup`.

Reply:
done, please take a look